### PR TITLE
Vendor the fused linear losses from Liger-Kernel into `trl.losses`

### PR DESCRIPTION
**Pull request by qgallouedec**
TRL's only import from `liger-kernel` is `liger_kernel.chunked_loss` (the fused linear DPO / KTO / GRPO / JSD losses). That module is pure PyTorch (chunked autograd + `torch.compile`, no Triton), and most of the GRPO and DPO variants in it were written for TRL. Upstream reviews have stalled (6 TRL PRs open since June, `chunked_loss/` untouched since 2026-06-29), so this PR brings the code home. It is a move, not a rewrite.

## What changes

- `trl/losses/`: the 4 losses and their 4 base classes, copied from Liger-Kernel v0.8.2 (`000be609`), BSD-2 notice kept, `Liger` prefix dropped: `FusedLinearDPOLoss`, `FusedLinearKTOLoss`, `FusedLinearGRPOLoss`, `FusedLinearJSDLoss`. Code is otherwise verbatim apart from lint (ruff formatting, `X | None` typing, `strict=True` on zips).
- Liger's unit tests for them: `tests/test_{dpo,kto,grpo,jsd}_loss.py` plus the `HFAlignmentLoss` / `HFDistillationLoss` references in `testing_utils.py`. The two tests that exercised `liger_kernel.transformers.grpo_loss` (Triton kernel, not vendored) are dropped.
- The 9 trainers (DPO, KTO, GRPO, Distillation, GKD, GOLD, SDPO, SDFT, IW-OPD) import the losses from `trl.losses` instead of `liger_kernel.chunked_loss`. 11 files, +28/-56, nothing else.

## What does not change

- No new config flag. `use_liger_kernel=True` keeps selecting the fused loss exactly as before, and still needs `liger-kernel` installed because transformers' `Trainer` patches the model kernels with it. The dependency goes away with the follow-ups.
- Numerics: same code, same results (bitwise identical to installed liger on random inputs for all loss types, fp32 and bf16). Docs unchanged.

## Tests

- `tests/test_{dpo,kto,grpo,jsd}_loss.py`: 1885 passed, 614 skipped (parameter combinations the losses reject), 8 workers on 1xH100, 5m54s. Compile-heavy; consider marking them `slow` if the CI lane gets too long.
- Trainer tests `-k liger` across the 9 trainers pass against `trl.losses`.

## Follow-ups

- Fold each trainer's fused path onto `_ChunkedLogProbFunction` and make it the default (GRPO drafted). This deletes `trl/losses/` and the second copy of the loss math, and closes the pre-existing gaps flagged in review (KTO reference under PEFT, KTO class weights, DPO `label_smoothing`, `chunk_size > 1` broadcast, Liger#1082).
- Deprecate `use_liger_kernel` in transformers (`use_kernels=True` already maps the same layers to `kernels-community/liger-kernels`), then drop the `liger` extra from TRL.

---

> [!NOTE]
> **Medium Risk**
> Changes the import path for fused training losses used across RLHF/distillation trainers; behavior is intended to be identical but any drift would affect optimization numerics.
> 
> **Overview**
> Moves the fused linear alignment and distillation losses out of `liger_kernel.chunked_loss` into a new **`trl.losses`** package (`FusedLinearDPOLoss`, `FusedLinearKTOLoss`, `FusedLinearGRPOLoss`, `FusedLinearJSDLoss`), adapted from Liger-Kernel v0.8.2 with the `Liger` prefix dropped.
> 
> Trainers and experiments that enable **`use_liger_kernel`** now construct these classes from **`trl.losses`** instead of importing Liger directly (GKD, GOLD, SDPO, SDFT, IW-OPD, GRPO, Distillation, plus the main DPO/KTO paths implied by the PR). Conditional `is_liger_kernel_available()` imports for the chunked loss module are removed where the fused JSD path is wired.
> 
> Adds Liger’s numerical parity tests (`tests/test_{dpo,kto,grpo,jsd}_loss.py`) and shared reference helpers in **`tests/testing_utils.py`** (`assert_verbose_allclose`, `HFAlignmentLoss`, `HFDistillationLoss`). Existing trainer tests are updated to assert `FusedLinear*` types from `trl.losses`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4bb3cf0f32e150d8b8c994f4d3aad2b15156ca28. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/losses/kto_loss.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+        preference_labels_chunk,
+        full_target,
+        ref_log_prob_chunk=None,
+        beta=0.1,
+        kl=None,
+    ):
+        """
+        Implements the Kahneman-Tversky Optimization (KTO) loss function.
+        Paper: "KTO: Model Alignment as Prospect Theory-Guided Optimization"
+        https://arxiv.org/abs/2402.01306
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/losses/kto_loss.py]: Fixed in c863fa78.

File: trl/losses/dpo_loss.py [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+            compute_nll_loss (bool): Whether to compute the NLL loss.
+            compiled (bool): Whether to use the torch compiled kernel.
+            use_ref_model (bool): Whether to use a reference model for the DPO loss.
+            average_log_prob (bool): Whether to average the log probability per non-masked token.
+            chunk_size (int): Size of chunks for processing.
+            loss_type (str): Variant of DPO loss to compute. One of:
+                "sigmoid", "hinge", "exo_pair", "nca_pair", "robust", "bco_pair",
+                "sppo_hard", "apo_zero", "apo_down", "discopop".
+            label_smoothing (float): Label smoothing for "robust" / "exo_pair" / cDPO.
+            discopop_tau (float): Temperature for the DiscoPOP modulation term.
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/losses/dpo_loss.py:302]: Kept verbatim from liger on purpose; these modules get folded into the trainers in a follow-up, docstrings will be converted then.

File: trl/experimental/zero_sync_grpo/zero_sync_grpo_trainer.py [hunk 3]
```diff
@@ -0,0 +1,1144 @@
...
+
+        # Compute per-token logprobs without ever materializing the [batch, seq, vocab] logits: the lm_head runs in
+        # chunks with an online logsumexp. Long completions make this the difference between training and an OOM.
+        # The patch replaces `forward`, and the generation engine decodes through that same object, so the standard
+        # forward is kept for calls without labels, which is what decoding does.
+        generation_forward = model.forward
+        # Under tensor parallelism the replicated lm_head would make every rank compute the whole vocabulary;
+        # passing the group makes each rank take every tp_size-th chunk instead, combined with two collectives.
+        tp_group = model._device_mesh.get_group() if args.tp_size > 1 else None
+        patch_chunked_lm_head(model, chunk_size=8192, temperature=self.temperature, tp_group=tp_group)
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/zero_sync_grpo/zero_sync_grpo_trainer.py]: Not part of this PR: the file was committed by mistake in 480abe6d, untracked in 18ac67ba.

File: trl/experimental/zero_sync_grpo/zero_sync_grpo_trainer.py [hunk 4]
```diff
@@ -0,0 +1,1144 @@
...
+            # option costs: the weights move during the step anyway, so the cache it throws away was already stale.
+            released = self._manager.release_memory()
+            self._metrics[mode]["generation/kv_released_gib"].append(released / 2**30)
+        if self._replica_group is not None:
+            samples = self._take_from_pool(num_samples)
+        else:
+            # Packing already removes the padding, and picking like-length samples would concentrate the long
+            # ones into a single batch, whose activations then blow past memory (a batch of 256 samples all near
+            # the length cap is over twice the tokens of a mixed one). Arrival order keeps the mix.
+            samples = [self._ready.popleft() for _ in range(num_samples)]
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/zero_sync_grpo/zero_sync_grpo_trainer.py]: Not part of this PR: the file was committed by mistake in 480abe6d, untracked in 18ac67ba.

File: trl/experimental/zero_sync_grpo/zero_sync_grpo_trainer.py [hunk 5]
```diff
@@ -0,0 +1,1144 @@
...
+        rewards = torch.stack([metrics["reward"] for metrics in rollout_metrics])
+        rewards_per_func = torch.stack([metrics["rewards_per_func"] for metrics in rollout_metrics])
+        for i, reward_func_name in enumerate(self.reward_func_names):
+            self._metrics[mode][f"rewards/{reward_func_name}/mean"].append(
+                torch.nanmean(rewards_per_func[:, i]).item()
+            )
+        self._metrics[mode]["reward"].append(torch.nanmean(rewards).item())
+        self._metrics[mode]["reward_std"].append(nanstd(rewards).item())
+        self._metrics[mode]["completions/mean_length"].append(
+            sum(metrics["completion_length"] for metrics in rollout_metrics) / len(rollout_metrics)
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/zero_sync_grpo/zero_sync_grpo_trainer.py]: Not part of this PR: the file was committed by mistake in 480abe6d, untracked in 18ac67ba.

File: trl/losses/kto_loss.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/losses/kto_loss.py]: Reverted in 7a902693: this PR keeps the vendored code as upstream except for lint reformatting; links get updated in the follow-up.

File: trl/trainer/kto_trainer.py [hunk 6]
```diff
@@ -983,9 +975,9 @@ def __init__(
                 )
             self.add_callback(SyncRefModelCallback(ref_model=self.ref_model, accelerator=self.accelerator))
 
-        # The Liger loss is built here, because it needs `self.ref_model`
-        if self.use_liger_kernel:
-            self.liger_loss = LigerFusedLinearKTOLoss(beta=self.beta, use_ref_model=(self.ref_model is not None))
+        # The fused linear loss is built here, because it needs `self.ref_model`
+        if self.use_fused_linear_loss:
+            self.fused_linear_loss = FusedLinearKTOLoss(beta=self.beta, use_ref_model=(self.ref_model is not None))
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/kto_trainer.py]: Pre-existing: identical to `use_liger_kernel=True` on main, this PR only moves the code. Goes away in the follow-up that folds the fused losses onto `_ChunkedLogProbFunction` (single loss implementation).

[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/kto_trainer.py]: Pre-existing: identical to `use_liger_kernel=True` on main, this PR only moves the code. Goes away in the follow-up that folds the fused losses onto `_ChunkedLogProbFunction` (single loss implementation).

File: trl/trainer/dpo_trainer.py [hunk 7]
```diff
@@ -793,62 +789,59 @@ def __init__(
...
-                        "wrong sequence. Use a weight-based adapter such as LoRA instead, or set "
-                        "`use_liger_kernel=False`."
+                        "`use_fused_linear_loss=True` is incompatible with prompt-learning PEFT methods "
+                        "(PromptTuning, PrefixTuning, P-Tuning). The fused linear DPO loss bypasses "
+                        "`PeftModel.forward()` by calling the backbone directly, so virtual tokens are never "
+                        "prepended and the loss is computed on the wrong sequence. Use a weight-based adapter such as "
+                        "LoRA instead, or set `use_fused_linear_loss=False`."
                     )
-            self.liger_loss = LigerFusedLinearDPOLoss(beta=args.beta, loss_type=self.loss_types[0])
+            self.fused_linear_loss = FusedLinearDPOLoss(beta=args.beta, loss_type=self.loss_types[0])
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/dpo_trainer.py]: Pre-existing: identical to `use_liger_kernel=True` on main, this PR only moves the code. Goes away in the follow-up that folds the fused losses onto `_ChunkedLogProbFunction` (single loss implementation).

File: trl/losses/fused_linear_unpaired_preference.py [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+        if average_log_prob:
+            log_probs = (per_token_logps_chunk * loss_mask_chunk).sum(-1) / loss_mask_chunk.sum(-1)
+        else:
+            log_probs = (per_token_logps_chunk * loss_mask_chunk).sum(-1)
+
+        chosen_logps_sum = (log_probs * preference_labels_chunk.unsqueeze(1)).sum()
+        rejected_logps_sum = (log_probs * (~preference_labels_chunk).unsqueeze(1)).sum()
+
+        chosen_logits_sum = (logits_chunk * preference_labels_chunk.unsqueeze(1)).sum()
+        rejected_logits_sum = (logits_chunk * (~preference_labels_chunk).unsqueeze(1)).sum()
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/losses/fused_linear_unpaired_preference.py:268]: Pre-existing: identical to `use_liger_kernel=True` on main, this PR only moves the code. Goes away in the follow-up that folds the fused losses onto `_ChunkedLogProbFunction` (single loss implementation).

**Review by kashif — APPROVED** [review]
